plugins/node/instrumentation-runtime-node/src/metrics/eventLoopDelayCollector.ts

```diff
@@ -0,0 +1,186 @@
+}
+
+export const metricNames: Record<
+  NodeJsEventLoopDelay,
+  { description: string }
+> = {
+  [NodeJsEventLoopDelay.min]: {
+    description: 'Event loop minimum delay.',
+  },
+  [NodeJsEventLoopDelay.max]: {
+    description: 'Event loop maximum delay.',
+  },
+  [NodeJsEventLoopDelay.mean]: {
+    description: 'Event loop mean delay.',
+  },
+  [NodeJsEventLoopDelay.stddev]: {
+    description: 'Event loop standard deviation delay.',
+  },
+  [NodeJsEventLoopDelay.p50]: {
+    description: 'Event loop 50 percentile delay.',
+  },
+  [NodeJsEventLoopDelay.p90]: {
+    description: 'Event loop 90 percentile delay.',
+  },
+  [NodeJsEventLoopDelay.p99]: {
+    description: 'Event loop 99 percentile delay.',
+  }
+};
+
+export interface EventLoopLagInformation {
+  min: number;
+  max: number;
+  mean: number;
+  stddev: number;
+  p50: number;
+  p90: number;
+  p99: number;
+}
+
+export class EventLoopDelayCollector extends BaseCollector<EventLoopLagInformation> {
+  private _histogram: IntervalHistogram;
+
+  constructor(
+    config: RuntimeNodeInstrumentationConfig = {},
+    namePrefix: string
+  ) {
+    super(config, namePrefix);
+    this._histogram = perf_hooks.monitorEventLoopDelay({
+      resolution: config.monitoringPrecision,
+    });
+  }
+
+  updateMetricInstruments(meter: Meter): void {
+    const delayMin = meter.createObservableGauge(
+      `${this.namePrefix}.${NodeJsEventLoopDelay.min}`,
+      {
+        description: metricNames[NodeJsEventLoopDelay.min].description,
+        unit: 's',
+      }
+    );
+    const delayMax = meter.createObservableGauge(
+      `${this.namePrefix}.${NodeJsEventLoopDelay.max}`,
+      {
+        description: metricNames[NodeJsEventLoopDelay.max].description,
+        unit: 's',
+      }
+    );
+    const delayMean = meter.createObservableGauge(
+      `${this.namePrefix}.${NodeJsEventLoopDelay.mean}`,
+      {
+        description: metricNames[NodeJsEventLoopDelay.mean].description,
+        unit: 's',
+      }
+    );
+    const delayStddev = meter.createObservableGauge(
+      `${this.namePrefix}.${NodeJsEventLoopDelay.stddev}`,
+      {
+        description: metricNames[NodeJsEventLoopDelay.stddev].description,
+        unit: 's',
+      }
+    );
+    const delayp50 = meter.createObservableGauge(
+      `${this.namePrefix}.${NodeJsEventLoopDelay.p50}`,
+      {
+        description: metricNames[NodeJsEventLoopDelay.p50].description,
+        unit: 's',
+      }
+    );
+    const delayp90 = meter.createObservableGauge(
+      `${this.namePrefix}.${NodeJsEventLoopDelay.p90}`,
+      {
+        description: metricNames[NodeJsEventLoopDelay.p90].description,
+        unit: 's',
+      }
+    );
+    const delayp99 = meter.createObservableGauge(
+      `${this.namePrefix}.${NodeJsEventLoopDelay.p99}`,
+      {
+        description: metricNames[NodeJsEventLoopDelay.p99].description,
+        unit: 's',
+      }
+    );
+
+    meter.addBatchObservableCallback(
```

Comment:
We have observed that this code produces inconsistent data in our production environment.

Currently as is the code scrapes the histogram once every "MONITORING_PRECISION" interval and adds it to a queue. It then exports a single entry from the queue based on the export interval of the metrics that you defined.

For example in our case we have a 1 minute export set up and a 5 second monitoring precision. So every minute 12 entries are added to the queue but we only export 1. This causes the queue to build up over time.

Additionally the sampling interval of once per 5 seconds for the event loop delay is not useful for capturing transient blips.

We modified the code to eliminate the scrape queue entirely, tweak the histogram to sample every 10 ms (this is the default setting in nodejs:perf_hooks and then export the summary metrics int he observable result. This resulted in much better data that we could use.

I can share the changes but they are pretty substantial so I'd be interested to hear back if I'm missing something in the current implementation.

Reply:
with your latest changes it is pretty much the same as what I have done locally. The only extra thing I did was reduce the interval for the event loop monitoring to the default value of 10ms. Just removing this line:
https://github.com/open-telemetry/opentelemetry-js-contrib/pull/2136/files#diff-9bee3da4de3a7ff765b777ba1c95c52a943f0dfc98eb5b3ecccd1830741bbeffR78

Perhaps this should still be configurable but given how this metric works a sampling interval of 5s is much less useful and unlikely to identify transient spikes I guess.

Reply:
I would also note that the minimum value reported by these metrics == the duration passed in on line 78. So if you do pass in 5s it will always report the lag as >= 5s which is pretty misleading. At least with a default value of 10ms you can measure lag < 5s

Reply:
Set default monitoringPrecision to 10ms

Reply:
I guess this conversation can be resolved right?